uniffi_bindgen/src/bindings/kotlin/templates/Helpers.kt

```diff
@@ -28,3 +28,73 @@ inline fun <T : FFIObject, R> T.use(block: (T) -> R) =
             // swallow
         }
     }
+
+internal typealias Handle = Long
+internal class ConcurrentHandleMap<T>(
+    private val leftMap: MutableMap<Handle, T> = mutableMapOf(),
+    private val rightMap: MutableMap<T, Handle> = mutableMapOf()
+) {
+    private val lock = java.util.concurrent.locks.ReentrantLock()
+    private val currentHandle = AtomicLong(0L)
+    private val stride = 1L
+
+    fun insert(obj: T): Handle =
+        lock.withLock {
+            rightMap[obj] ?:
+                currentHandle.getAndAccumulate(stride) { a, b -> a + b }
+                    .also { handle ->
+                        leftMap[handle] = obj
+                        rightMap[obj] = handle
+                    }
+            }
+
+    fun <R> callWithResult(handle: Handle, fn: (T) -> R): R =
+        lock.withLock {
+            val obj = leftMap[handle] ?: throw RuntimeException("Panic: handle not in handlemap")
+            fn.invoke(obj)
+        }
+
+    fun get(handle: Handle) = lock.withLock {
+        leftMap[handle]
+    }
+
+    fun delete(handle: Handle) { 
+        this.remove(handle)
+    }
+
+    fun remove(handle: Handle): T? =
+        lock.withLock {
+            leftMap.remove(handle)?.let { obj ->
+                rightMap.remove(obj)
+                obj
+            }
+        }
+}
+
+interface ForeignCallback : com.sun.jna.Callback {
+    public fun invoke(handle: Long, method: Int, args: RustBuffer.ByValue): RustBuffer.ByValue
+}
+
+internal abstract class CallbackInternals<CallbackInterface>(
+    val foreignCallback: ForeignCallback
+) {
+    val handleMap = ConcurrentHandleMap<CallbackInterface>()
+
+    abstract fun register(lib: _UniFFILib)
+
+    fun drop(handle: Long): RustBuffer.ByValue {
+        return handleMap.remove(handle).let { RustBuffer.ByValue() }
+    }
+
+    fun lift(n: Long) = handleMap.get(n)
+
+    fun read(buf: ByteBuffer) = lift(buf.getLong())
```

Comment:
As with the Rust side of things, I'm kind of dubious about `lift`/`read`/`write` here, since I don't think we can guarantee that the handle is still good.

Reply:
> I don't think we can guarantee that the handle is still good.

I'm not sure I understand why it's not true. I'll have to think about this.

Reply:
Revisiting this, I'm not sure I recall exactly what my concern was. I guess the other implicit link in the chain of argument here is that we don't ever `read` or `write` as part of some long-term serialization that might keep a reference to the handle around after it's been freed, we only do them as part of a transient shuffling of data from one side of the FFI to the other. So...I think I'm OK with this having concrete `read` and `write` implementations 👍